Doc design blob UUID & upload process (#150)
dandi/dandi-api: doc/design/blob-uuid-1.md

A blob is referenced by a key generation function. the value of this key is used to index the blob in the database table of identities and on the s3 storage. `s3://dandiarchive/blob/<3digitskey>/<next3digitskey>/<restofkey>`

> ** database table of identities **: indexed by the key and contains columns for as many types of checksums as we want. these checksums constitute identities of the blob and should use functions that will always return the same value.

Reviewer: @dchiquito - shouldn't this be indexed by the UUID with the ETag being one of the identities? only because two ETag's could collide (in theory, the probably for that is higher than the probability for UUID4s to collide on the same system)

> ** database table of identities **: indexed by the key and contains columns for as many types of checksums as we want. these checksums constitute identities of the blob and should use functions that will always return the same value.

Author: if we are to have dedicated columns for checksums, I thought they all could just go into "blobs" table with an index, UUID (unless UUID to be used as index, but I thought it is inferior), ETag (which we might want to name 'dandietag' or alike to make more explicit), sha256 columns to start with, i.e. that there would be no need in another table with 1-to-1 relationship to blobs table.  or what am I missing?

> ** database table of identities **: indexed by the key and contains columns for as many types of checksums as we want. these checksums constitute identities of the blob and should use functions that will always return the same value.

Reviewer: i'm fine with `index, uuid, dandietag, sha256, blake3, <future checksums>` whether the index is built in or uuid, i don't care a whole lot. but there may be performance penalties for search, which would likely be given a checksum, find uuid, or given a uuid, find checksums.

**key generation function**: the current suggestion (from satra) for this function is UUID4.

**DANDI ETag generation function**: this is an agreement between API server and DANDI CLI for now to ensure there is an algorithm that consistently returns the correct number of parts given a `contentSize` in bytes. this should in theory also be the AWS ETag generation function.

## Upload (from a CLI)

### Check process (for de-duplication):
1. CLI sends expected digest with digest type (e.g., ETag)
2. API checks for existence of ETag in identity table.
3. if exists, returns key, if not returns an error.

### Blob upload process:

> 1. CLI initiates upload sending contentSize + ETag
> 2. API generates key, determines upload location `s3://dandiarchive/blob/<3digitskey>/<next3digitskey>/<key>`
> 4. API saves initial data (size, ETAG, key) to Upload table
> 3. API returns presigned URLs

Author: ```suggestion
1. CLI `/uploads/initiate` <-- `{contentSize , digests: { "dandi-s3-etag": "..." }`. API:
   1. generates key, determines upload location `s3://dandiarchive/blob/<3digitskey>/<next3digitskey>/<key>`
   2. saves initial data (size, ETAG, key) to Upload table
   3. API returns presigned URLs
```

4. CLI verifies that the number+size of parts match the calculated number of parts (i.e. server and CLI are using the same ETag generation function)
5. CLI uploads to presigned URLs, and for each part checks ETag on return. Any part upload failure can be retried without involving API.

> 6. CLI sends part info to API, API responds with presigned completion URL

Author: ```suggestion
6. `/uploads/complete/`: CLI sends parts info to API, API responds with presigned completion URL
```

7. CLI completes upload and checks final ETag. Mismatch aborts the upload.
8. API validates the size and that the initially reported ETag matches the actual ETag
9. Check for collision, since some other task could have finished by this time with the same object

> 10. API adds key + ETag to Asset table, delete from Upload table

Reviewer 2: ```suggestion
10. API adds key + ETag to AssetBlob table, delete from Upload table
```

> 11. API kicks off background process to calculate checksums

Author: ```suggestion
8. CLI `/uploads/validate` <-- `uploads:upload_id` --> `blob:uuid`. API
    1. validates the size and that the initially reported ETag matches the actual ETag
    2. checks for collision, since some other task could have finished by this time with the same object
    3. adds key + ETag to AssetBlob table, delete from Upload table
    4. kicks off background process to calculate checksums
```

> 8. API validates the size and that the initially reported ETag matches the actual ETag
> 9. Check for collision, since some other task could have finished by this time with the same object
> 10. API adds key + ETag to AssetBlob table, delete from Upload table
> 11. API kicks off background process to calculate checksums

Reviewer 2: How about this:
```suggestion
8. CLI `/uploads/validate` <-- `{contentSize , digests: { "dandi-s3-etag": "..." }`. API
    1. Checks for an existing AssetBlob. If it exists, no need to validate: return the key (UUID).
    2. Looks up the Upload from the DB with the given ETag. If not present, return 404.
    3. Verify that the object described by the Upload exists, is the right size, and has the right ETag. If not, return 400.
    4. Adds key + ETag + size to AssetBlob table and delete the Upload.
    4. kicks off background process to calculate checksums
    5. Return key (UUID) of AssetBlob so CLI can do `POST .../assets/` to mint new asset
```
The big difference is that the ETag is used to lookup the Upload, and there is some more explicit checking for existing info in the DB.

> 11. API kicks off background process to calculate checksums

Author: > Checks for an existing AssetBlob. If it exists, no need to validate 

if so -- still would need to remove an upload (so 2.) from uploads.  But matching could lead to upload of others:

> here is some more explicit checking for existing info in the DB.

how is it "more explicit" if it is not by an `upload_id`?  matching by metadata could match uploads of others

A blob is referenced by a key generation function. the value of this key is used to index the blob in the database table of identities and on the s3 storage. `s3://dandiarchive/blob/<3digitskey>/<next3digitskey>/<restofkey>`

** database table of identities **: indexed by the key and contains columns for as many types of checksums as we want. these checksums constitute identities of the blob and should use functions that will always return the same value.

**key generation function**: the current suggestion (from satra) for this function is UUID4.

**DANDI ETag generation function**: this is an agreement between API server and DANDI CLI for now to ensure there is an algorithm that consistently returns the correct number of parts given a `contentSize` in bytes. this should in theory also be the AWS ETag generation function.

## Upload (from a CLI)

### Check process (for de-duplication):
1. CLI sends expected digest with digest type (e.g., ETag)
2. API checks for existence of ETag in identity table.
3. if exists, returns key, if not returns an error.

### Blob upload process:

> 1. CLI initiates upload sending contentSize + ETag

Author: currently we do
```
            resp = self.post(
                "/uploads/initialize/",
                json={
                    "file_name": f"{dandiset_id}/{version_id}/{asset_path}",
                    "file_size": total_size,
                },
            )
```
so already conflating "pure blob" upload with an asset upload (has a file_name which could be different across blobs).  So I wonder if uploads should just be initiated from  `dandiset/.../versions/draft` end point?

> 1. CLI initiates upload sending contentSize + ETag

Reviewer: that is weird. i can see filename without all those other things being useful, since that can be put on the s3 object metadata for content-disposition, but what is the full file_name being used for in the upload process (dandiset/version seems completely unnecessary)

> 1. CLI initiates upload sending contentSize + ETag

Reviewer 2: `file_name` is actually unused by the API, it was required at some point but has been removed. Apparently it was never removed from the CLI, and having extra JSON fields doesn't cause any problems, the data is simply ignored. That line can be safely removed, it is meaningless.

> 1. CLI initiates upload sending contentSize + ETag

Reviewer 2: I think planning for different upload pipelines at this point is not productive. For now we have a well defined problem, uploading a file from a client machine using the CLI. If we want to support different ways to ingest files in the future, we can build a different upload flow at that time.

My understanding is that to get to release faster, we were going to just trust the CLI that content is valid for now. Handling metadata and validating during initialization would make things substantially more complicated. Is this a migration blocker, or can we leave it out for now @yarikoptic ?

> 1. CLI initiates upload sending contentSize + ETag

Reviewer: i think @yarikoptic - simply meant sending the info in the json header in that post (and not really full metadata validation). but i can also see sending contentsize/digest info in the call to finalize upload, since that's where validation would happen. 

and i do agree that making that json match with the subset of assetmetadata keys would probably be a good thing, but not essential at this point.

so the payload on `/uploads/validate` the upload would be something like this.

```json
{ 
  "identifier": uuid,
  "contentSize":  12345,
  "digest":  {"dandi:dandi-s3-etag": "aadd-123"}
}
```

this formulation still needs a change in the cli.

> 1. CLI initiates upload sending contentSize + ETag

Reviewer 2: I think `identifier` not be included, or at least be optional:
```
{ 
  "contentSize":  12345,
  "digest":  {"dandi:dandi-s3-etag": "aadd-123"}
}
```
When the CLI needs to upload file, the first step is to calculate some digest (in this case ETag) and ask the API using this endpoint if it already exists; there's no way for the CLI to know the UUID.
Since we already need to support that use case, I don't see what specifying `identifier` gains us.

> 1. CLI initiates upload sending contentSize + ETag

Author: FWIW re `file_path` and content-disposition: I would stand that blobs have no filename (IMHO UUID is better than a wrong/misleading/disclosing one from the first ever uploaded copy of the blob):
- Pre-signed urls produced from `*/assets/{uuid}/download` endpoint can have appropriate for the asset content-disposition
- git-annex doesn't care
- "published" manifests would have proper paths and thus filenames for each of the blobs

So indeed, no need to provide file path or name anywhere here

> 1. CLI initiates upload sending contentSize + ETag

Reviewer: > Pre-signed urls produced from */assets/{uuid}/download endpoint can have appropriate for the asset content-disposition

has this been tested. let me add it as an issue.

> 1. CLI initiates upload sending contentSize + ETag

Reviewer 2: @yarikoptic I agree with returning the blob ID only from `/uploads/validation/`, but I don't see the need for an intermediate `upload-id`. To support deduplication, `/uploads/validate/` needs to accept a request body like:
```
{ 
  "contentSize":  12345,
  "digest":  {"dandi:dandi-s3-etag": "aadd-123"}
}
```
Since the ETag is already there (and is already being relied on to uniquely identify blobs), we can just look up the upload using that. I don't see what including an `upload-id` gains us.

> 1. CLI initiates upload sending contentSize + ETag

Reviewer: what if two people are uploading the same blob?

> 1. CLI initiates upload sending contentSize + ETag

Reviewer: would you randomly choose the first uuid?

> 1. CLI initiates upload sending contentSize + ETag

Reviewer 2: To have a collision, the second person would need to call `/uploads/initialize/` with the same ETag as the first person. This request should fail because there is already an upload.

> 1. CLI initiates upload sending contentSize + ETag

Reviewer: i thought we allowed for parallel uploads of the same content (since we are using UUID, and the ETag check does not happen on the upload table) - which if we check with ETag during upload stage while nice, would make the person A wait for person B to finish uploading. this would also ensure that the fastest uploaded blob gains precedence.

> 1. CLI initiates upload sending contentSize + ETag

Author: > I don't see what including an upload-id gains us

- a clear identification of the upload in question, and avoidance of matching by metadata provided in `/upload/initialize/` and thus avoiding problems with duplicate uploads as @satra has mentioned - we can just allow them to happen and the first one to complete would be the one the `key` of which would be used for that blob
- anything about upload (cancel etc) should be done via its `upload_id` instead of matching uploads and possibly canceling some other parallel one (ATM we do not have `DELETE` on uploads but I guess we should get one eventually)
That is why below I would prefer [my](https://github.com/dandi/dandi-api/pull/150#discussion_r591868008) over [@dchiquito ](https://github.com/dandi/dandi-api/pull/150#discussion_r591902088) proposal below.

> 1. CLI initiates upload sending contentSize + ETag

Reviewer 2: OK, I think at this point we're really overloading `/upload/initialize/`. How about breaking it into two endpoints:
 * `GET /blobs/` checks for an existing asset with an arbitrary digest (but ETag for now):
 ```
 { 
  "contentSize":  12345,
  "digest":  {"dandi:dandi-s3-etag": "aadd-123"}
}
```
* `POST /uploads/validate/{upload_id}/` starts validation for an upload, identified by upload ID.

2. API generates key, determines upload location `s3://dandiarchive/blob/<3digitskey>/<next3digitskey>/<key>`
4. API saves initial data (size, ETAG, key) to Upload table
3. API returns presigned URLs
4. CLI verifies that the number+size of parts match the calculated number of parts (i.e. server and CLI are using the same ETag generation function)

> 5. CLI uploads to presigned URL and checks part ETags on return. Any failures can be retried without involving API.

Author: ```suggestion
5. CLI uploads to presigned URLs, and for each part provides Content-MD5 header field  and checks ETag on return. Any part upload failure can be retried without involving API.
```

Reviewer 2: The API server needs to specify all headers when presigning the request, so to provide the `Content-MD5` header instead of checking the resulting ETag would require the CLI to pass all of the part MD5s to the API when initializing the upload.

Reviewer: that is correct. i would suggest that this step only involve the CLI doing the check without sending content-MD5 to aws.

Author: Right, forgot about that, adjusted my suggestion

6. CLI sends part info to API, API responds with presigned completion URL
7. CLI completes upload and checks final ETag. Mismatch aborts the upload.
8. API validates the size and that the initially reported ETag matches the actual ETag
9. Check for collision, since some other task could have finished by this time with the same object

> 10. API adds key + ETag to Asset table, delete from Upload table

Author: in 8.-10. results in a return a record with blob UUID (either a new one or "reused"), right?

Reviewer 2: My thought was that ETag would be a good "identity" key for blobs, since all blobs necessarily have an ETag, and ETag has to be used to check if a blob already exists. We could use UUID instead, but it might get confusing managing blob UUIDs and asset UUIDs.

Reviewer: i would think so.

Reviewer: @dchiquito - what's the Asset Table here? there is no asset in this upload process.

Reviewer 2: My mistake, I mean AssetBlob. At this point the data is in S3 and is known to be valid, so it needs to be canonicalized in the DB.

> 11. API kicks off background process to calculate checksums

Author: a side-note that if we stay true to current design -- addition of any new checksum should change asset UUID unless we keep those outside of the metadata records and treat them as some "dynamic" overlay ...

Reviewer 2: I think dynamically injecting the checksums like we already do with asset path info would be better. I can imagine someone uploading some new data to a dandiset, checking that it's there in the UI, then publishing it immediately before the background checksum calculations complete.

Reviewer: any new checksum is an overlay on the uuid - should not change uuid of a blob.

A blob is referenced by a key generation function. the value of this key is used to index the blob in the database table of identities and on the s3 storage. `s3://dandiarchive/blob/<3digitskey>/<next3digitskey>/<restofkey>`

** database table of identities **: indexed by the key and contains columns for as many types of checksums as we want. these checksums constitute identities of the blob and should use functions that will always return the same value.

**key generation function**: the current suggestion (from satra) for this function is UUID4.

**DANDI ETag generation function**: this is an agreement between API server and DANDI CLI for now to ensure there is an algorithm that consistently returns the correct number of parts given a `contentSize` in bytes. this should in theory also be the AWS ETag generation function.

## Upload (from a CLI)

### Check process (for de-duplication):

> 1. CLI sends expected digest with digest type (e.g., ETag)
> 2. API checks for existence of ETag in identity table.
> 3. if exists, returns key, if not returns an error.

Reviewer 2: ```suggestion
1. CLI `GET /blobs/` <-- expected digest with digest type (e.g., ETag). API:
    1. Looks for an existing AssetBlob with the given digest(s).
    2. If it exists, return the key (UUID).
    3. It it doesn't exist, return 404.
2. If 404, start upload process.
3. if key, call `POST .../assets/` with the key to register the new asset.
```

Author: That is pretty much how it is now -- a responsibility of a good client to explicitly check first. 

I think we should leave above "Check" possibility as is for its own life, and make `/uploads/initialize` with similar request as to check, and do check internally (thus returning e.g. HTTP 409 Conflict with a Location or some other header field to point to existing blob key).  This would preclude "bad" clients to not check first, cause traffic, and us only later returning them a key for an already existing blob at `validate`.  And that is why I liked it when you (I believe) added to have `/uploads/initialize` to get the size and digest.

Reviewer 2: @yarikoptic That would work, but it doesn't resolve the overloading of HTTP actions. I liked the logical separation of having either
`GET /blobs/` -> `POST .../assets/` (no upload involved, so no calls to `/upload/...`)
or
`GET /blobs/` -> `POST /upload/initialize+complete+validate/` -> `POST .../assets/`
That keeps the upload process as a more distinct pathway.

This is just splitting hairs over nitty gritty HTTP design, though. I can implement it your way if you feel strongly.

Author: "strongly" would be the wrong word since I am not an experienced API designer ;)  
but it feels that it could help us avoid unnecessary waste of bandwidth so feels "more right" to me.

Note that a client could still do `GET /blobs/` first to proceed current/first path, but going directly to `POST /upload/initialize` (and handling a possible error indicating presence of the blob) would avoid that additional initial `GET /blobs/` call completely

Reviewer 2: How about adding the `GET /blobs/` endpoint so clients have a dedicated, reliable way to get look up existing assets, and making `/upload/initialize/` simply fail when attempting to upload a digest that's already registered?

Author: Yes, that is what [I had in mind above as well](https://github.com/dandi/dandi-api/pull/150#discussion_r592744335): `/blobs` is generally good to have, but not needed for "upload logic". Just that `/upload/initialize` should "simply fail" informatively - provide `blobs:key` in the return headers so that client could proceed with it to ```9. CLI calls `POST .../assets/` with the key to register the new asset.```
Note that it does leave possibility for a race condition (GC removes blob:key before CLI calls POST it for a new asset) but I guess it would be very unlikely... may be also there could be some timestamp increment on `blobs` record for that `key` so GC does not even consider blobs which were recently "shown interest in"? Is there something like "access time" for DB records or overhead would be too high?

Reviewer 2: I think we're on the same page, I'm committing this suggestion.

I think race conditions related to the garbage collector are garbage collector problems, which are not in the scope of this design doc.

   1. generates key, determines upload location `s3://dandiarchive/blob/<3digitskey>/<next3digitskey>/<key>`
   2. saves initial data (size, ETag, key) to Upload table
   3. API returns presigned URLs
4. CLI verifies that the number+size of parts match the calculated number of parts (i.e. server and CLI are using the same ETag generation function)
5. CLI uploads to presigned URLs, and for each part checks ETag on return. Any part upload failure can be retried without involving API.
6. `/uploads/complete/`: CLI sends parts info to API, API responds with presigned completion URL
7. CLI completes upload and checks final ETag. Mismatch aborts the upload.
8. CLI `/uploads/validate` <-- `uploads:upload_id` --> `blobs:key`. API
    1. validates the size and that the initially reported ETag matches the actual ETag
    2. checks for collision, since some other task could have finished by this time with the same object
        1. no collision - adds key + ETag to AssetBlob table, delete from Upload table.
        2. there is collision - delete key record and file in blob store, will return key of an existing record

>     4. kicks off background process to calculate checksums
> 

Reviewer 2: ```suggestion
    4. kicks off background process to calculate checksums
9. CLI calls `POST .../assets/` with the key to register the new asset.

```

## Garbage Collection process

> 1. if dangling key's are found, API does garbage collection on keys that are older than 24 hours.

Author: ```suggestion
1. if dangling keys (not used by any asset) are found, API does garbage collection on keys that are older than 24 hours.
```

> A blob is referenced by a key generation function. the value of this key is used to index the blob in the database table of identities and on the s3 storage. `s3://dandiarchive/blob/<3digitskey>/<next3digitskey>/<restofkey>`

Author: ```suggestion
A blob is referenced by a key generation function. the value of this key is used to index the blob in the database table of identities and on the s3 storage. `s3://dandiarchive/blob/<3digitskey>/<next3digitskey>/<key>`
```
this would be consistent with the form used below in upload 1.1, and the form I would generally prefer even though at the little cost of duplicating first 6 digits in the directory names and the key name.  I find such a form easier to reuse later on and possibly avoid the file name to start with `-` if it happens to be the 7th digit in the key.

Author: but I do not feel too strong about it (I do not immediately see where using `restofkey` would really fail) so feel free to reject but then adjust path below in upload 1.1 to match.

** database table of identities **: indexed by the key and contains columns for as many types of checksums as we want. these checksums constitute identities of the blob and should use functions that will always return the same value.

**key generation function**: the current suggestion (from satra) for this function is UUID4.

**DANDI ETag generation function**: this is an agreement between API server and DANDI CLI for now to ensure there is an algorithm that consistently returns the correct number of parts given a `contentSize` in bytes. this should in theory also be the AWS ETag generation function.

## Upload (from a CLI)

### Check process (for de-duplication):
1. CLI sends expected digest with digest type (e.g., ETag)
2. API checks for existence of ETag in identity table.
3. if exists, returns key, if not returns an error.

### Blob upload process:
1. CLI `/uploads/initialize` <-- `{contentSize , digests: { "dandi-s3-etag": "..." }`. API:
   1. generates key, determines upload location `s3://dandiarchive/blob/<3digitskey>/<next3digitskey>/<key>`
   2. saves initial data (size, ETag, key) to Upload table

>    3. API returns presigned URLs

Author: ```suggestion
   3. API returns `uploads:upload_id` and presigned URLs
```

4. CLI verifies that the number+size of parts match the calculated number of parts (i.e. server and CLI are using the same ETag generation function)
5. CLI uploads to presigned URLs, and for each part checks ETag on return. Any part upload failure can be retried without involving API.
6. `/uploads/complete/`: CLI sends parts info to API, API responds with presigned completion URL
7. CLI completes upload and checks final ETag. Mismatch aborts the upload.
8. CLI `/uploads/validate` <-- `uploads:upload_id` --> `blobs:key`. API
    1. validates the size and that the initially reported ETag matches the actual ETag
    2. checks for collision, since some other task could have finished by this time with the same object
        1. no collision - adds key + ETag to AssetBlob table, delete from Upload table.

>         2. there is collision - delete key record and file in blob store, will return key of an existing record

Author: ```suggestion
        2. there is collision - delete key record and file in uploads, will return key of an existing record
```

** database table of identities **: indexed by the key and contains columns for as many types of checksums as we want. these checksums constitute identities of the blob and should use functions that will always return the same value.

**key generation function**: the current suggestion (from satra) for this function is UUID4.

**DANDI ETag generation function**: this is an agreement between API server and DANDI CLI for now to ensure there is an algorithm that consistently returns the correct number of parts given a `contentSize` in bytes. this should in theory also be the AWS ETag generation function.

## Upload (from a CLI)

### Check process (for de-duplication):
1. CLI `GET /blobs/` <-- expected digest with digest type (e.g., ETag). API:
    1. Looks for an existing AssetBlob with the given digest(s).
    2. If it exists, return the key (UUID).
    3. It it doesn't exist, return 404.
2. If 404, start upload process.
3. if key, call `POST .../assets/` with the key to register the new asset.

### Blob upload process:
1. CLI `/uploads/initialize` <-- `{contentSize , digests: { "dandi-s3-etag": "..." }`. API:

>    1. generates key, determines upload location `s3://dandiarchive/blob/<3digitskey>/<next3digitskey>/<key>`
>    2. saves initial data (size, ETag, key) to Upload table
>    3. API returns `uploads:upload_id` and presigned URLs

Reviewer 2: ```suggestion
   1. Checks if there is already an asset blob with the given digest. If so, return 409 CONFLICT.
   2. generates key, determines upload location `s3://dandiarchive/blob/<3digitskey>/<next3digitskey>/<key>`
   3. saves initial data (size, ETag, key) to Upload table
   4. API returns `uploads:upload_id` and presigned URLs
```

A blob is referenced by a key generation function. the value of this key is used to index the blob in the database table of identities and on the s3 storage. `s3://dandiarchive/blob/<3digitskey>/<next3digitskey>/<key>`

** database table of identities **: indexed by the key and contains columns for as many types of checksums as we want. these checksums constitute identities of the blob and should use functions that will always return the same value.

**key generation function**: the current suggestion (from satra) for this function is UUID4.

**DANDI ETag generation function**: this is an agreement between API server and DANDI CLI for now to ensure there is an algorithm that consistently returns the correct number of parts given a `contentSize` in bytes. this should in theory also be the AWS ETag generation function.

## Upload (from a CLI)

### Check process (for de-duplication):

> 1. CLI `POST /blobs/digests/` <-- `{"algorithm": "dandi:dandi-s3-etag", "value": "123etag123"}`. API:

Reviewer 2: ```suggestion
1. CLI `POST /blobs/digests/` <-- `{"algorithm": "dandi:dandi-etag", "value": "123etag123"}`. API:
```

    1. Looks for an existing AssetBlob with the given digest(s).

>     2. If it exists, return the key (UUID).

Reviewer 2: ```suggestion
    2. If it exists, return the `blobs:blob_id`.
```

    3. It it doesn't exist, return 404.
2. If 404, start upload process.

> 3. if key, call `POST .../assets/` with the key to register the new asset.

Reviewer 2: ```suggestion
3. if `blob_id`, call `POST .../assets/` with the `blob_id` to register the new asset.
```

### Blob upload process:

> 1. CLI `/uploads/initialize` <-- `{contentSize , digests: { "dandi-s3-etag": "..." }`. API:

Reviewer: my only comment here is that we will want `contentSize` here and in the remaining places, since that's in the asset metadata. also the current digest doesn't look like above, but hopefully we will make that adjustment soon.

> 1. CLI `/uploads/initialize` <-- `{contentSize , digests: { "dandi-s3-etag": "..." }`. API:

Reviewer 2: `file_size` is here, which is a synonym for `contentSize`.
I typo'd the `algorithm`, it should be fixed now.

> 1. CLI `/uploads/initialize` <-- `{contentSize , digests: { "dandi-s3-etag": "..." }`. API:

Author: we have contentSize in the schema, blobs are data blobs so might even have no file (at least with a name) so I would stick with `contentSize` without adding synonyms.

> 1. CLI `/uploads/initialize` <-- `{"content_size": 12345 , "digest": {"algorithm": "dandi:dandi-s3-etag", "value": "..."}`. API:

Reviewer 2: ```suggestion
1. CLI `/uploads/initialize` <-- `{"content_size": 12345 , "digest": {"algorithm": "dandi:dandi-etag", "value": "..."}`. API:
```

>    1. Checks if there is already an asset blob with the given digest. If so, return 409 CONFLICT.

Author: ```suggestion
   1. Checks if there is already an asset blob with the given digest. If so, return 409 CONFLICT with a key in the header (`Location`?).
```

>    1. Checks if there is already an asset blob with the given digest. If so, return 409 CONFLICT with the `blobs:key` in the header (`Location`?).

Author: @dchiquito so is it to be `Location` header field in response and thus we should remove `?` or something else?

Reviewer 2: I implemented it to use the `Location` header, but I'm happy to leave the `?` until the CLI half is implemented in case it turns out something else would be preferable.

>    1. Checks if there is already an asset blob with the given digest. If so, return 409 CONFLICT with the `blobs:key` in the header (`Location`?).

Reviewer 2: ```suggestion
   1. Checks if there is already an asset blob with the given digest. If so, return 409 CONFLICT with the `blobs:blob_id` in the header (`Location`?).
```

   2. generates key, determines upload location `s3://dandiarchive/blob/<3digitskey>/<next3digitskey>/<key>`
   3. saves initial data (size, ETag, key) to Upload table
   4. API returns `uploads:upload_id` and presigned URLs
4. CLI verifies that the number+size of parts match the calculated number of parts (i.e. server and CLI are using the same ETag generation function)

> 5. CLI uploads to presigned URLs, and for each part checks ETag on return. Any part upload failure can be retried without involving API.

Reviewer 3: What exactly is the client supposed to check the ETag against?  The md5 digest of the part?

Author: yes AFAIK

6. `/uploads/{upload_id}/complete/`: CLI sends parts info to API, API responds with presigned completion URL

> 7. CLI completes upload and checks final ETag. Mismatch aborts the upload.

Author: what does "aborts the upload" would entail -- removing `uploads:upload_id` record and target `s3://dandiarchive/blob/<3digitskey>/<next3digitskey>/<key>` or something else?

Reviewer 2: I think if the CLI wants to abort at any point, it should just stop. The user could kill the CLI at any point, or the CLI could run into some random CLI bug. It's the server's job to have lifecycle rules in place to clean up stale multipart uploads that were not completed, and garbage collection to clean up uploads that were never minted into Assets.

> 7. CLI completes upload and checks final ETag. Mismatch aborts the upload.

Reviewer 3: "checks final ETag" — Where is this ETag that is being checked?

Author: after you POST (IIRC) to that "completion URL" you should receive final ETag, and it should correspond to the `dandi-etag` computed before initiating upload

> 8. CLI `/uploads/validate` <-- `uploads:upload_id` --> `blobs:key`. API

Reviewer 2: ```suggestion
8. CLI `POST /uploads/validations/{uploads:upload_id}/` --> `blobs:key`. API
```

Author: Why? Do you expect a table of validations to be also queried for some kind of status?

We have `/initialize` and `/complete`  actions, and from the description below action should be atomic enough to not need any `validations` table.  For background processes I think we did talk about `/jobs` where checksums calculation could be a type of job associated with a `key`.

Reviewer 2: The endpoint basically takes an upload ID, looks up the Upload object, and verifies that the upload was completed correctly by the client. 
I see now that `POST /uploads/validations/{uploads:upload_id}/` is misleading, I will change my suggestion to `POST /uploads/validate/{uploads:upload_id}/`.
As an alternative, it could also be `POST /uploads/validate/` <-- `{"uuid": "1234...5678"}`, but I like that less than putting the ID in the URL.

Author: > As an alternative, it could also be POST /uploads/validate/ <-- {"uuid": "1234...5678"}, but I like that less than putting the ID in the URL.

I think `POST /uploads/validate/ <-- {"uploads:upload_id": "1234...5678"}` (note: not  `uuid` AKA key which isn't yet returned) would keep it more consistent with other actions like `/initialize` and `/complete`, and consistency would be good since otherwise need to remember where an id goes into URL and where into the load for some imperative `action/` endpoint.  For direct operations like `GET` or `DELETE /uploads/{upload_uuid}` it indeed makes a good sense to have it in a URL, while again `PUT` or `POST` suggest a data record to be sent, so why then in URL?...  could then even be `GET /uploads/validate/{uploads:upload_id}/` and that would be really odd ;-)

Having said that, if was `/uploads/initialize -> upload_id`, `/uploads/{upload_id}/complete/` and `/uploads/{upload_id}/validate/` - I might have not complained since it would have paralleled better with `.../assets` endpoints

Author: I like this `/uploads/{upload_uuid}{,/<operation>}` idea!

> 6. `/uploads/complete/`: CLI sends parts info to API, API responds with presigned completion URL
> 7. CLI completes upload and checks final ETag. Mismatch aborts the upload.
> 8. CLI `/uploads/validate` <-- `uploads:upload_id` --> `blobs:key`. API

Reviewer 2: ```suggestion
6. `/uploads/{upload_id}/complete/`: CLI sends parts info to API, API responds with presigned completion URL
7. CLI completes upload and checks final ETag. Mismatch aborts the upload.
8. CLI `/uploads/{upload_id}/validate/` <-- `uploads:upload_id` --> `blobs:key`. API
```

> 8. CLI `/uploads/{upload_id}/validate/` <-- `uploads:upload_id` --> `blobs:key`. API

Author: ```suggestion
8. CLI `GET /uploads/{upload_id}/validate/` --> `blobs:key`. API
```
or should be some POST?

Reviewer 2: It should be POST, as it is doing things, not just returning information.

Author: adjusted suggestion with an explicitly empty record to POST

> 8. CLI `/uploads/{upload_id}/validate/` <-- `uploads:upload_id` --> `blobs:key`. API

Reviewer 2: ```suggestion
8. CLI `/uploads/{upload_id}/validate/` <-- `uploads:upload_id` --> `blobs:blob_id`. API
```

> 8. CLI `POST /uploads/{upload_id}/validate/` <-- `{}` --> `blobs:key`. API

Reviewer 2: ```suggestion
8. CLI `POST /uploads/{upload_id}/validate/` <-- `{}` --> `blobs:blob_id`. API
```

    1. validates the size and that the initially reported ETag matches the actual ETag
    2. checks for collision, since some other task could have finished by this time with the same object
        1. no collision - adds key + ETag to AssetBlob table, delete from Upload table.

>         2. there is collision - delete key record and file in uploads, will return key of an existing record

Author: ```suggestion
        2. there is collision - deletes `blobs:key` record, `uploads:upload_id` record, and `s3://dandiarchive/blob/<3digitskey>/<next3digitskey>/<key>`, will return key of an existing record
```

>         2. there is collision - deletes `blobs:key` record, `uploads:upload_id` record, and `s3://dandiarchive/blob/<3digitskey>/<next3digitskey>/<key>`, will return key of an existing record

Reviewer 2: ```suggestion
        2. there is collision - deletes `blobs:blob_id` record, `uploads:upload_id` record, and `s3://dandiarchive/blob/<3digitskey>/<next3digitskey>/<key>`, will return key of an existing record
```

    4. kicks off background process to calculate checksums

> 9. CLI calls `POST .../assets/` with the key to register the new asset.

Reviewer 2: ```suggestion
9. CLI calls `POST .../assets/` with the `blob_id` to register the new asset.
```
